src/bvals/bvals.hpp

```diff
@@ -92,6 +92,62 @@ class BoundaryBase {
   static bool called_;
 };
 
+//----------------------------------------------------------------------------------------
+//! \class BoundarySwarms
+//  \brief centralized class for interacting with each individual swarm boundary data
+class BoundarySwarms : public BoundaryBase, BoundaryCommunication {
+ public:
+  BoundarySwarms(std::weak_ptr<MeshBlock> pmb, BoundaryFlag *input_bcs,
+                 ParameterInput *pin);
+
+  // variable-length arrays of references to all BoundarySwarm instances
+  std::vector<std::shared_ptr<BoundarySwarm>> bswarms;
+
+  void SetBoundaryFlags(BoundaryFlag bc_flag[]) {
+    for (int i = 0; i < 6; i++) {
+      bc_flag[i] = block_bcs[i];
+    }
+  }
+
+  // inherited functions (interface shared with BoundaryVariable objects):
+  // ------
+  // called before time-stepper:
+  void SetupPersistentMPI() final; // setup MPI requests
+
+  // called before and during time-stepper (currently do nothing for swarms):
+  void StartReceiving(BoundaryCommSubset phase) final {}
+  void ClearBoundary(BoundaryCommSubset phase) final {}
+
+  int AdvanceCounterPhysID(int num_phys) { return 0; }
```

Comment:
Does this do anything anywhere in the code? Maybe we should remove it from the base class.

Reply:
It doesn't seem to do anything anywhere, which I don't understand... as far as I can tell the `bvars_next_phys_id_` quantity isn't getting updated (in swarms OR in CC bvars), which as far as I can tell it needs to be for MPI tags to work properly. So I think I'm missing something... this PR should probably not get merged until I figure this out (or someone tells me what's going on).

It's also possible that `bvars_next_phys_id_` needs to be moved down to e.g. `BoundaryBase` and not be separate between `BoundarySwarms` and `BoundaryValues`

Reply:
Hmm ok... We may want to merge this MR in before figuring this out if addressing this is part of a larger boundaries refactor.

I'm not sure any of us have looked deeply at that part of the code.